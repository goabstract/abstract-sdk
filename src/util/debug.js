// @flow
import debug from "debug";

export const log = debug("abstract");
