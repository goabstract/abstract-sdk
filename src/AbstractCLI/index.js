// @flow
import path from "path";
import { spawn } from "child_process";
import { Buffer } from "buffer";
import find from "lodash/find";
import locatePath from "locate-path";
import JSONStream from "JSONStream";
import { objectFileDescriptor } from "../utils";
import { log } from "../debug";
import type {
  AbstractInterface,
  ProjectDescriptor,
  CommitDescriptor,
  BranchDescriptor,
  PageDescriptor,
  FileDescriptor,
  LayerDescriptor,
  CollectionDescriptor,
  AccessTokenOption
} from "../types";
import { throwCLIError } from "../errors";

const logSpawn = log.extend("AbstractCLI:spawn");
const logStdoutError = log.extend("AbstractCLI:stdout:error");
const logStdoutData = log.extend("AbstractCLI:stdout:data");
const logStderrData = log.extend("AbstractCLI:stderr");
const logClose = log.extend("AbstractCLI:close");

function parsePath(input: ?string): ?Array<string> {
  if (!input) return;
  return input.split(path.delimiter || ":");
}

export type Options = {
  accessToken: AccessTokenOption,
  cliPath?: string[],
  apiUrl?: string,
  cwd?: string
};

export default class AbstractCLI implements AbstractInterface {
  _optionAccessToken: AccessTokenOption;
  cliPath: string;
  apiUrl: string;
  cwd: string;

  constructor({
    cwd = process.cwd(),
    accessToken,
    cliPath = parsePath(process.env.ABSTRACT_CLI_PATH) || [
      // Relative to cwd
      path.join(cwd, "abstract-cli"),
      // Relative to node_modules in cwd (also makes test easier)
      path.join(
        cwd,
        "node_modules/@elasticprojects/abstract-cli/bin/abstract-cli"
      ),
      // macOS App
      "/Applications/Abstract.app/Contents/Resources/app.asar.unpacked/node_modules/@elasticprojects/abstract-cli/bin/abstract-cli"
    ],
    apiUrl = "https://api.goabstract.com"
  }: Options = {}) {
    this.cwd = cwd;
    this._optionAccessToken = accessToken;
    this.apiUrl = apiUrl;

    try {
      this.cliPath = path.resolve(cwd, locatePath.sync(cliPath));
    } catch (error) {
      throw new Error(`Cannot find abstract-cli in "${cliPath.join(":")}"`);
    }
  }

  accessToken = async () =>
    typeof this._optionAccessToken === "function"
      ? this._optionAccessToken()
      : this._optionAccessToken;

  async spawn(args: string[]) {
    const accessToken = await this.accessToken();
    const userTokenArgs =
      typeof accessToken === "string" ? ["--user-token", accessToken] : [];

    return new Promise((resolve, reject) => {
      const spawnArgs = [
        this.cliPath,
        [
          ...userTokenArgs,
          "--api-url",
          this.apiUrl,
          ...args // First args win for https://github.com/spf13/cobra
        ],
        { cwd: this.cwd }
      ];

      logSpawn(spawnArgs);
      const abstractCli = spawn(...spawnArgs);

      let stderrBuffer = Buffer.from("");
      abstractCli.stderr.on("data", chunk => {
        logStderrData(chunk.toString());
        stderrBuffer = Buffer.concat([stderrBuffer, chunk]);
      });

      const stream = JSONStream.parse();
      stream.on("data", data => {
        logStdoutData(data);
        resolve(data);
      });

      abstractCli.stdout
        .on("data", data => {
          stream.write(data);
        })
        .on("error", error => {
          logStdoutError(error.toString());
          reject(error);
        })
        .on("end", () => {
          stream.end();
        });

      abstractCli.on("error", reject);
      abstractCli.on("close", errorCode => {
        logClose(errorCode);

        if (errorCode !== 0) {
          const response = JSON.parse(stderrBuffer.toString());
          try {
            throwCLIError(response, spawnArgs[0], { ...(spawnArgs[1]: any) });
          } catch (error) {
            reject(error);
          }
        }
      });
    });
  }

  async resolveDescriptor<T: *>(objectDescriptor: T): Promise<T> {
    if (objectDescriptor.sha !== "latest") return objectDescriptor;

    const commits = await this.commits.list(objectDescriptor, { limit: 1 });

    try {
      return {
        ...objectDescriptor,
        sha: commits[0].sha
      };
    } catch (error) {
      throw new Error(
        `Could not resolve sha "latest" for ${JSON.stringify(objectDescriptor)}`
      );
    }
  }

  commits = {
    list: async (
      objectDescriptor:
        | BranchDescriptor
        | FileDescriptor
        | PageDescriptor
        | LayerDescriptor,
      options?: { limit?: number } = {}
    ) => {
      const fileIdArgs = objectDescriptor.fileId
        ? ["--file-id", objectDescriptor.fileId]
        : [];

      const layerIdArgs = objectDescriptor.layerId
        ? ["--layer-id", objectDescriptor.layerId]
        : [];

      const limitArgs = options.limit
        ? ["--limit", options.limit.toString()]
        : [];

      const data = await this.spawn([
        "commits",
        objectDescriptor.projectId,
        objectDescriptor.branchId,
        ...fileIdArgs,
        ...layerIdArgs,
        ...limitArgs
      ]);

      return data.commits;
    },
    info: async (
      objectDescriptor:
        | BranchDescriptor
        | FileDescriptor
        | CommitDescriptor
        | LayerDescriptor
    ) => {
      // Avoid using resolveDescriptor to prevent extra request
      if (objectDescriptor.sha === "latest") {
        const commits = await this.commits.list(objectDescriptor, { limit: 1 });
        return commits[0];
      } else {
        const data = await this.spawn([
          "commit",
          objectDescriptor.projectId,
          objectDescriptor.sha
        ]);

        return data.commit;
      }
    }
  };

  changesets = {
    info: async (commitDescriptor: CommitDescriptor) => {
      return this.spawn([
        "changeset",
        commitDescriptor.projectId,
        "--commit",
        commitDescriptor.sha,
        "--branch",
        commitDescriptor.branchId
      ]);
    }
  };

  files = {
    list: async (branchDescriptor: BranchDescriptor) => {
      branchDescriptor = await this.resolveDescriptor(branchDescriptor);

      const data = await this.spawn([
        "files",
        branchDescriptor.projectId,
        branchDescriptor.sha
      ]);

      return data.files;
    },
    info: async (fileDescriptor: FileDescriptor) => {
      fileDescriptor = await this.resolveDescriptor(fileDescriptor);

      const data = await this.spawn([
        "file",
        fileDescriptor.projectId,
        fileDescriptor.sha,
        fileDescriptor.fileId
      ]);

      const file = {
        ...data.file,
        _pages: data.pages
      };

      return file;
    }
  };

  pages = {
    list: async (fileDescriptor: FileDescriptor) => {
      const file: File = await this.files.info(fileDescriptor);
      // $FlowFixMe: _pages not in type
      return file._pages;
    },
    info: async (pageDescriptor: PageDescriptor) => {
      const pages = await this.pages.list(objectFileDescriptor(pageDescriptor));
      return find(pages, { id: pageDescriptor.pageId });
    }
  };

  layers = {
    list: async (fileDescriptor: FileDescriptor) => {
      fileDescriptor = await this.resolveDescriptor(fileDescriptor);

      return this.spawn([
        "layers",
        fileDescriptor.projectId,
        fileDescriptor.sha,
        fileDescriptor.fileId
      ]);
    },
    info: async (layerDescriptor: LayerDescriptor) => {
      layerDescriptor = await this.resolveDescriptor(layerDescriptor);

      const data = await this.spawn([
        "layer",
        "meta",
        layerDescriptor.projectId,
        layerDescriptor.sha,
        layerDescriptor.fileId,
        layerDescriptor.layerId
      ]);

      return data.layer;
    }
  };

  data = {
    info: async (layerDescriptor: LayerDescriptor) => {
      layerDescriptor = await this.resolveDescriptor(layerDescriptor);

      return this.spawn([
        "layer",
        "data",
        layerDescriptor.projectId,
        layerDescriptor.sha,
        layerDescriptor.fileId,
        layerDescriptor.layerId
      ]);
    }
  };

  collections = {
    list: (projectOrBranchDescriptor: ProjectDescriptor | BranchDescriptor) => {
      const branchArgs = projectOrBranchDescriptor.branchId
        ? ["--branch", projectOrBranchDescriptor.branchId]
        : [];

      return this.spawn([
        "collections",
        projectOrBranchDescriptor.projectId,
        ...branchArgs
      ]);
    },
    info: (collectionDescriptor: CollectionDescriptor) => {
      return this.spawn([
        "collection",
        collectionDescriptor.projectId,
        collectionDescriptor.collectionId
      ]);
    }
  };

  branches = {
    list: async (
      projectDescriptor: ProjectDescriptor,
      options: { filter?: "active" | "archived" | "mine" } = {}
    ) => {
      const branchesArgs = options.filter ? ["--filter", options.filter] : [];

      const data = await this.spawn([
        "branches",
        projectDescriptor.projectId,
        ...branchesArgs
      ]);
      return data.branches;
    },
    info: async (branchDescriptor: BranchDescriptor) => {
      return await this.spawn([
        "branch",
        "load",
        branchDescriptor.branchId,
        branchDescriptor.projectId
      ]);
    }
  };
}
