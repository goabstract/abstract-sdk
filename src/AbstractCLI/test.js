// @flow
import child_process from "child_process";
import path from "path";
import { Readable } from "readable-stream";
import get from "lodash/get";
import {
  buildOptions,
  buildProjectDescriptor,
  buildCommitDescriptor,
  buildBranchDescriptor,
  buildFileDescriptor,
  buildPageDescriptor,
  buildLayerDescriptor,
  buildCollectionDescriptor
} from "../support/factories";
import { log } from "../debug";
import AbstractCLI from "./";

jest.mock("child_process");

const logTest = log.extend("AbstractCLI:test");

function buildTextStream(text?: string): ReadableStream {
  const stream = new Readable();
  stream._read = () => {}; // required interface

  if (text !== undefined) {
    stream.push(text);
    stream.push(null); // end of stream
  }

  return stream;
}

const responses = {
  commits: {
    list: () => ({
      stdout: JSON.stringify({
        commits: [{ sha: "commit-sha" }, { sha: "next-commit-sha" }]
      })
    })
  },
  files: {
    list: () => ({
      stdout: JSON.stringify({
        files: [{ id: "file-id" }, { id: "not-file-id" }]
      })
    }),
    info: () => ({
      stdout: JSON.stringify({ file: { id: "file-id" } })
    })
  },
  pages: {
    info: () => ({
      stdout: JSON.stringify({
        pages: [{ id: "not-page-id" }, { id: "page-id" }],
        file: {} // required
      })
    })
  },
  layers: {
    info: () => ({
      stdout: JSON.stringify({ layer: { id: "layer-id" } })
    })
  },
  branches: {
    list: () => ({
      stdout: JSON.stringify({
        branches: [{ id: "foo" }, { id: "bar" }]
      })
    }),
    info: () => ({
      stdout: JSON.stringify({ id: "baz" })
    })
  }
};

describe(AbstractCLI, () => {
  test("throws when abstract-cli cannot be located", () => {
    expect(
      () =>
        new AbstractCLI(
          buildOptions({
            cliPath: ["./fixtures/missing/abstract-cli"]
          })
        ).cliPath
    ).toThrow('Cannot find abstract-cli in "./fixtures/missing/abstract-cli"');
  });

  describe("without module cache to reset process.env", () => {
    beforeEach(() => {
      jest.resetModules();
    });

    test.each([
      [
        // Relative path
        "./fixtures/abstract-cli",
        "fixtures/abstract-cli"
      ],
      [
        // Ignores misssing abstract-cli path
        "/missing/abstract-cli:./fixtures/abstract-cli",
        "fixtures/abstract-cli"
      ],
      [
        // Prefers first matched abstract-cli path
        "./fixtures/other/abstract-cli/:fixtures/abstract-cli",
        "fixtures/other/abstract-cli"
      ]
    ])(
      "configures options.cliPath from process.env.ABSTRACT_CLI_PATH=%p",
      (cliPath, expected) => {
        process.env.ABSTRACT_CLI_PATH = cliPath;
        const { default: AbstractCLI } = require("./");
        expect(new AbstractCLI(buildOptions()).cliPath).toBe(
          path.resolve(expected)
        );
      }
    );

    test("prefers options.cliPath=%p over process.env.ABSTRACT_CLI_PATH", () => {
      process.env.ABSTRACT_CLI_PATH = "./fixtures/other/abstract-cli";
      const { default: AbstractCLI } = require("./");
      expect(
        new AbstractCLI(
          buildOptions({
            cliPath: ["./fixtures/abstract-cli"]
          })
        ).cliPath
      ).toBe(path.resolve("fixtures/abstract-cli"));
    });
  });

  describe("with mocked child_process.spawn", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test.each([
      // collections
      ["collections.list", buildProjectDescriptor()],
      ["collections.list", buildBranchDescriptor()],
      ["collections.info", buildCollectionDescriptor()],
      // commits
      ["commits.list", buildBranchDescriptor()],
      ["commits.list", buildFileDescriptor()],
      ["commits.list", buildLayerDescriptor()],
      ["commits.info", buildCommitDescriptor()],
      // changesets
      ["changesets.info", buildCommitDescriptor()],
      // files
      [
        "files.info",
        buildFileDescriptor(),
        {
          responses: [responses.files.info()],
          result: { id: "file-id" }
        }
      ],
      [
        "files.list",
        buildBranchDescriptor(),
        {
          responses: [responses.files.list()],
          result: [{ id: "file-id" }, { id: "not-file-id" }]
        }
      ],
      [
        "files.info",
        buildFileDescriptor({ sha: "latest" }),
        { responses: [responses.commits.list()] }
      ],
      [
        "files.list",
        buildCommitDescriptor(),
        {
          responses: [responses.files.list()],
          result: [{ id: "file-id" }, { id: "not-file-id" }]
        }
      ],
      [
        "files.list",
        buildBranchDescriptor({ sha: "latest" }),
        { responses: [responses.commits.list()] }
      ],
      // pages
      ["pages.list", buildFileDescriptor()],
      [
        "pages.info",
        buildPageDescriptor(),
        {
          responses: [responses.pages.info()],
          result: { id: "page-id" }
        }
      ],
      // layers
      ["layers.list", buildFileDescriptor()],
      [
        "layers.list",
        buildFileDescriptor({ sha: "latest" }),
        { responses: [responses.commits.list()] }
      ],
      [
        "layers.info",
        buildLayerDescriptor(),
        {
          responses: [responses.layers.info()],
          result: { id: "layer-id" }
        }
      ],
      [
        "layers.info",
        buildLayerDescriptor({ sha: "latest" }),
        {
          responses: [responses.commits.list(), responses.layers.info()],
          result: { id: "layer-id" }
        }
      ],
      // data
      ["data.info", buildLayerDescriptor()],
      [
        "data.info",
        buildLayerDescriptor({ sha: "latest" }),
        { responses: [responses.commits.list()] }
      ],
      // branches
      [
        "branches.list",
        buildProjectDescriptor(),
        { responses: [responses.branches.list()] }
      ],
      [
        "branches.info",
        buildBranchDescriptor(),
        { responses: [responses.branches.info()] }
      ]
    ])("%s(%p)", async (property, args, options = {}) => {
      args = Array.isArray(args) ? args : [args];
      logTest(property, args);

      const transport = new AbstractCLI(
        buildOptions({ cliPath: ["./fixtures/abstract-cli"] })
      );

      const transportMethod = get(transport, property).bind(transport);

      if (options.responses) {
        options.responses.forEach(response => {
          child_process.spawn.mockReturnValueOnce({
            stdout: buildTextStream(response.stdout),
            stderr: buildTextStream(response.stderr),
            on: jest.fn().mockReturnThis()
          });
        });
      }

      child_process.spawn.mockReturnValueOnce({
        stdout: buildTextStream("{}"),
        stderr: buildTextStream(""),
        on: jest.fn().mockReturnThis()
      });

      const result = transportMethod(...args);
      await expect(await result).resolves;

      if (options.result) {
        expect(await result).toEqual(options.result);
      }

      expect({ spawn: child_process.spawn.mock.calls }).toMatchSnapshot();
    });
  });
});
