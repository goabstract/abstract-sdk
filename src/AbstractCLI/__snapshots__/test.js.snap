// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractCLI with mocked child_process.spawn branches.info({"branchId": "branch-id", "projectId": "project-id", "sha": "branch-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "branch",
        "load",
        "branch-id",
        "project-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn branches.list({"projectId": "project-id"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "branches",
        "project-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn changesets.info({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "changeset",
        "project-id",
        "--commit",
        "commit-sha",
        "--branch",
        "branch-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn collections.info({"branchId": "branch-id", "collectionId": "collection-id", "projectId": "project-id"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "collection",
        "project-id",
        "collection-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn collections.list({"branchId": "branch-id", "projectId": "project-id", "sha": "branch-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "collections",
        "project-id",
        "--branch",
        "branch-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn collections.list({"projectId": "project-id"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "collections",
        "project-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commit",
        "project-id",
        "commit-sha",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
        "--layer-id",
        "layer-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "file-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "projectId": "project-id", "sha": "branch-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn data.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "latest"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
        "--layer-id",
        "layer-id",
        "--limit",
        "1",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layer",
        "data",
        "project-id",
        "commit-sha",
        "file-id",
        "layer-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn data.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layer",
        "data",
        "project-id",
        "layer-sha",
        "file-id",
        "layer-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn files.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "file-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "file",
        "project-id",
        "file-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn files.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "latest"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
        "--limit",
        "1",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "file",
        "project-id",
        "commit-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "branch-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "files",
        "project-id",
        "branch-sha",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "files",
        "project-id",
        "commit-sha",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "latest"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--limit",
        "1",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "files",
        "project-id",
        "commit-sha",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn layers.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "latest"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
        "--layer-id",
        "layer-id",
        "--limit",
        "1",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layer",
        "meta",
        "project-id",
        "commit-sha",
        "file-id",
        "layer-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn layers.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layer",
        "meta",
        "project-id",
        "layer-sha",
        "file-id",
        "layer-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn layers.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "file-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layers",
        "project-id",
        "file-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn layers.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "latest"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "commits",
        "project-id",
        "branch-id",
        "--file-id",
        "file-id",
        "--limit",
        "1",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "layers",
        "project-id",
        "commit-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn pages.info({"branchId": "branch-id", "fileId": "file-id", "pageId": "page-id", "projectId": "project-id", "sha": "page-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "file",
        "project-id",
        "page-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;

exports[`AbstractCLI with mocked child_process.spawn pages.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "file-sha"}) 1`] = `
Object {
  "spawn": Array [
    Array [
      "<PROJECT_ROOT>/fixtures/abstract-cli",
      Array [
        "--user-token",
        "abstract-token",
        "--api-url",
        "https://api.goabstract.com",
        "file",
        "project-id",
        "file-sha",
        "file-id",
      ],
      Object {
        "cwd": "<PROJECT_ROOT>",
      },
    ],
  ],
}
`;
